Feature/6207 config single file

sagarafr opened the pull request:
Changelog: Feature: Allow ``conan config install`` of a single file
Docs: https://github.com/conan-io/docs/pull/1908

Close https://github.com/conan-io/conan/issues/6207

* Add the possibility to specify a configuration for the command `conan config install`
* Copy only the configuration files in the directory, for the command `conan config install dirname`

Question :
* For what purpose, do we need the files copy, when the directory is pass by parameter ?

- [X] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [ ] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

mgappa committed:
feat(client): Specify configuration file installation

* In order to specify a file configuration for the command `conan config install`,
we test if the file is a zip file first. If it is not a zip, then we use the same
logic in the function _process_folder, by using the name file.
* For some convinance, we have comment the all other files copy during made
by the command `conan config install`

mgappa committed:
fix(pep8): Fixed pep8 intendation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=7840) <br/>All committers have signed the CLA.

memsharded commented on conans/client/conf/config_installer.py:
```diff
@@ -86,6 +86,45 @@ def _filecopy(src, filename, dst):
+    elif filename == "remotes.json":
+        # Fix for Conan 2.0
+        raise ConanException("remotes.json install is not supported yet. Use 'remotes.txt'")
+    # TODO : do we need this ? in which case we need to copy all the folder source
```
yes, please leave this as-is, it is a feature to copy any other file in the folder, not just the known ones. For example, it is possible to install python files and use them in recipes (not recommended, legacy, but possible)

sagarafr replied:
OK. I was not sure of the usage, and, in the issue, the user want no necessary file copy.

memsharded commented on conans/client/tools/files.py:
```diff
@@ -101,41 +101,42 @@ def print_progress(the_size, uncomp_size):
-                except Exception as e:
-                    output.error("Error extract %s\n%s" % (file_.filename, str(e)))
-        output.writeln("")
+    if zipfile.is_zipfile(filename):
```
I don't understand this, this code is specifically for zip files, so no need to check it. 
At most it could be a "precondition" check that checks it and raise a meaningful error otherwise, but that would be a different feature (that will require its own unit test)

sagarafr replied:
When I use to run the command `conan configure install` on a file, the file is categorized as a file, then, try to unzip this file and this raise an exception, because it's not a ZIP file.
My meaning, in this change, is to check if the file is a ZIP file, first, to avoid unnecessary exception and after that threat the file as a real file.
I can change that in order to first check if the input file is a configuration, and then, run the unzip function in order to keep the raising error. Or I can keep this change and add a meaning full error output. What is your preference ?

memsharded reviewed (changes requested):
Hi @sagarafr

Thanks very much for contributing this PR. Please have a look:

- the tests are broken. This shouldn' t be breaking existing tests (https://conan-ci.jfrog.info/blue/organizations/jenkins/ConanTestSuite/detail/PR-7840/1/pipeline)
- It would be necessary to add some test that covers the new functionality. Please ask for help if needed
- Please sign the CLA above (sometimes it is an issue of using the right user while creating commits, using the same user as github one)

sagarafr committed:
chore(tools): Implement is_compressed_file

* is_compressed_file permit to test if a file is a zip, or a gzip,
or a targz file
* Remove the check is_zipfile in unzip function, in order to
test if a file is a zip in the ./client/conf/config_installer.py file

sagarafr committed:
chore(conf): Keep custom files install and specify compressed files

* Keep custom files installation
* Test if the uri is a compressed file and use decompression
  configuration function
* Test if the uri is a regular file and use file configuration function

memsharded commented on conans/client/conf/config_installer.py:
```diff
@@ -201,9 +207,12 @@ def _process_config(config, cache, output, requester):
+        elif config.type == "compressed":
             with tmp_config_install_folder(cache) as tmp_folder:
                 _process_zip_file(config, config.uri, cache, output, tmp_folder)
+        elif config.type == "file":
```
Hi @sagarafr

I am having a look, and there is a problem: we cannot change existing behavior unless it is declared a bug. At the moment the type for a zipped file, the type is "file", and that cannot be changed. It is not good enough to introduce a ``compressed`` type for differentiating this. The current ``file`` behavior has to be respected and maintained.

So I think it would be better to try to process files capturing the exception when the decompression fails.

sagarafr replied:
Indeed, I understand my mistake when I have push. I just try to find the right place to make the check in order to have a cleaner code and have all tests pass.

sagarafr committed:
fix(tools): Fixed is_compressed_file function

* Add first, if the filename is a file in order to remove
the `compressed` state

sagarafr committed:
chore(conf): Remove `compressed` state

* Remove the `compressed` state, by checking if a `file` type is a zip or a tar* or a gzip

sagarafr committed:
refactor(conf): Remove commented code

sagarafr committed:
chore(tools): Open a file first in is_compressed_file

* In order to raise an exception if a file exist or not, open a file in
  order to check if it is a gzip file first, then check if it is a zip
or a tar* file

sagarafr committed:
test(functional): Add a test for file installation

* Test the command conan config install with a specific file

sagarafr committed:
fix(test): Use mkdtemp instead of TemporaryDirectory

memsharded reviewed (approved):
Good, thanks very much for contributing this feature!

mgappa commented:
Thanks you for your patience

sagarafr commented:
Thanks your for your patience
